Provide hostname to GSASL, enabling GSSAPI authentication

jelmer opened the pull request:
This pull request is a follow-up to https://github.com/dino/dino/pull/517, and makes Dino provide the FQDN of the server it is talking to to GSASL.

This enables GSSAPI (Kerberos) authentication, and fixes #482. I've been using this with prosody for a while.

jelmer committed:
Import libgsasl vapi from vala-extra-vapis.

jelmer committed:
Use gsasl sha1 functions.

jelmer committed:
Use gsasl for SASL operations.

This simplifies the SASL module somewhat, and brings in support for
more SASL mechanisms that are supported by GSASL:
https://www.gnu.org/software/gsasl/

jelmer committed:
Keep track of the remote hostname and pass it to sasl.

This enables authentication with GSSAPI (e.g. Kerberos). Fixes #482

mar-v-in commented on xmpp-vala/src/core/xmpp_stream.vala:
```diff
@@ -57,10 +58,14 @@ public class XmppStream {
             }
             if (stream == null) {
-                stream = yield (new SocketClient()).connect_async(new NetworkService("xmpp-client", "tcp", this.remote_name.to_string()));
+                GLibFixes.Resolver resolver = GLibFixes.Resolver.get_default();
```
This originally was the fallback case for when there is no SRV record. I guess your change will break the fallback and make SRV records required.

jelmer replied:
My understanding is that the original code would only work with SRV as well, based on reading the documentation for NetworkService: https://valadoc.org/gio-2.0/GLib.NetworkService.html

Wouldn't this need to use NetworkAddress rather than NetworkService to use a plain hostname rather than a SRV record?

jelmer replied:
I'm marking this as resolved since I've reverted this part of the code - whether this is doing the right thing is orthogonal.

mar-v-in commented on xmpp-vala/src/module/sasl.vala:
```diff
@@ -162,60 +74,34 @@ namespace Xmpp.Sasl {
-            } else {
-                stderr.printf("No supported mechanism provided by server at %s\n", stream.remote_name.to_string());
+            flag.sasl_session.set_property(Gsasl.Property.AUTHID, name);
+            flag.sasl_session.set_property(Gsasl.Property.HOSTNAME, stream.remote_hostname);
```
According to XEP 233, the hostname for GSSAPI is provided with the SASL mechanisms, see https://xmpp.org/extensions/xep-0233.html#examples
Can we use this instead of passing the hostname around everywhere?

jelmer replied:
I've updated this to use the hostname from the SASL mechanism, and reverted the passing around of hostnames everywhere.

mar-v-in commented on xmpp-vala/src/module/sasl.vala:
```diff
@@ -44,110 +39,27 @@ namespace Xmpp.Sasl {
+            if (node.ns_uri != NS_URI)
+                return;
+            if (node.name == "success") {
+                Flag flag = stream.get_flag(Flag.IDENTITY);
```
I think with the new logic, the server signature (that is sent with the success nonza) is not verified any more (as described in points 10 and 11 in https://wiki.xmpp.org/web/SASL_and_SCRAM-SHA-1). Can you check how to verify it through gsasl?

jelmer replied:
Done.

mar-v-in reviewed (commented):
I gave some comments inline.

I generally would love if we had a way to make the libgsasl dependency optional through a plugin, but it is probably to much work right now (especially as we want to store the scram salted password instead of cleartext password longterm). We can still split it out in a plugin later on, after we defined some proper APIs for that.

jelmer committed:
Retrieve the hostname from the mechanisms XML element, per XEP-0233.

On e.g. Prosody this requires enabling the saslname module:
https://modules.prosody.im/mod_saslname.html

jelmer committed:
Don't mark authentication as successful until the local SASL session has successfully terminated too.

This ensures that e.g. server signature in SCRAM-SHA1 is successfully verified.

jelmer commented:
Please take another look, I believe I have addressed all review comments.

Neustradamus commented:
@mar-v-in: Any news for this PR?
